### PR TITLE
 add numstr and printstr support

### PR DESCRIPTION
**Pull request by fleg**
Like I promised

### CODE REVIEW

File: lib/asn1/decoders/der.js [hunk 1]
```diff
@@ -133,6 +133,8 @@ DERNode.prototype._decodeStr = function decodeStr(buffer, tag) {
     return { unused: unused, data: buffer.raw() };
   } else if (tag === 'ia5str' || tag === 'utf8str') {
     return buffer.raw().toString();
+  } else if(tag === 'numstr' || tag === 'printstr') {
+    return buffer.raw().toString('ascii');
```
Comment (indutny) [inline lib/asn1/decoders/der.js]: I think it should validate it here.

File: lib/asn1/encoders/der.js [hunk 2]
```diff
@@ -61,21 +61,38 @@ DERNode.prototype._encodeComposite = function encodeComposite(tag,
...
     }
     return this._createEncoderBuffer(buf);
+  } else if (tag === 'numstr') {
+    if (!/^[1-9 ]*$/.test(str))
+      return this.reporter.error('Encoding of string type: numstr supports ' +
+                                 'only digits and space');
+
+    return this._createEncoderBuffer(str);
+  } else if (tag === 'printstr') {
+    if (!/^[A-Za-z0-9 '\(\)\+,\-\.\/:=\?]*$/.test(str))
```
Comment (indutny) [inline lib/asn1/encoders/der.js]: Please add braces

File: lib/asn1/encoders/der.js [hunk 3]
```diff
@@ -61,21 +61,38 @@ DERNode.prototype._encodeComposite = function encodeComposite(tag,
...
     return this._createEncoderBuffer(str);
-  else if (tag === 'bmpstr') {
+  } else if (tag === 'bmpstr') {
     var buf = new Buffer(str.length * 2);
     for (var i = 0; i < str.length; i++) {
       buf.writeUInt16BE(str.charCodeAt(i), i * 2);
     }
     return this._createEncoderBuffer(buf);
+  } else if (tag === 'numstr') {
+    if (!/^[1-9 ]*$/.test(str))
```
Comment (indutny) [inline lib/asn1/encoders/der.js]: Please add braces.

File: test/ping-pong-test.js [hunk 4]
```diff
@@ -45,6 +45,14 @@ describe('asn1.js ping/pong', function() {
       this.bmpstr();
     }, 'hello');
 
+    test('numstr', function() {
+      this.numstr();
+    }, '1234 5678');
+
+    test('printstr', function() {
+      this.bmpstr();
```
Comment (indutny) [inline test/ping-pong-test.js]: `printstr`?

File: lib/asn1/decoders/der.js (hunk 1 above)
Comment (fleg) [inline lib/asn1/decoders/der.js]: ok, maybe add some methods for validation in Node class?
something like this:
```javascript
Node.prototype._isNumstr = function(str) {
  return /^[1-9 ]*$/.test(str);
};
```

Comment (indutny) [inline lib/asn1/decoders/der.js]: Actually, it makes sense. Good idea. But please name the `function isNumstr`.

**Comment by fleg** [comment]
fixed

**Comment by indutny** [comment]
LGTM, thank you!

**Comment by indutny** [comment]
Landed and released, thanks again!

**Pull request closed without merging**